src/Microsoft.OData.Core/Json/JsonReaderExtensions.cs

```diff
@@ -315,6 +316,40 @@ internal static ODataValue ReadAsUntypedOrNullValue(this IJsonReader jsonReader)
             };
         }
 
+        internal static ODataValue ReadODataValue(this IJsonReader jsonReader)
+        {
+            if (jsonReader.NodeType == JsonNodeType.PrimitiveValue)
+            {
+                object primitiveValue = jsonReader.ReadPrimitiveValue();
+
+                return primitiveValue.ToODataValue();
+            }
+            else if (jsonReader.NodeType == JsonNodeType.StartObject)
+            {
+                jsonReader.ReadStartObject();
+                ODataResourceValue resourceValue = new ODataResourceValue();
+                IList<ODataProperty> propertyList = new List<ODataProperty>();
+
+                while (jsonReader.NodeType != JsonNodeType.EndObject)
+                {
+                    ODataProperty property = new ODataProperty();
+                    property.Name = jsonReader.ReadPropertyName();
+                    property.Value = jsonReader.ReadODataValue();
+                    propertyList.Add(property);

```

Comment:
what about nested objects? Should this call into ReadODataValue, and have a separate case for if (jsonReader.NodeType == JsonNodeType.Property)?